Add support to `time` and `time64` datatypes

kavirajk opened the pull request:
## Summary
<!-- A short description of the changes with a link to an open issue. -->

This is built on top of [other PR ](https://github.com/ClickHouse/clickhouse-go/pull/1596) preserving original authors commits.


## Checklist
Delete items not relevant to your PR:
- [ ] Unit and integration tests covering the common scenarios were added
- [ ] A human-readable description of the changes was provided to include in CHANGELOG
- [ ] For significant changes, documentation in https://github.com/ClickHouse/clickhouse-docs was updated with further explanations or tutorials

shivanshuraj1333 committed:
feat: time and time64 column support

shivanshuraj1333 committed:
feat: add time and time64 support to codegen and column_gen.go

shivanshuraj1333 committed:
test: add tests for time and time64 with setup and protocol wrappers

shivanshuraj1333 committed:
chore: document time and time64 go type mappings in types.md

SpencerTorres committed:
Merge branch 'main' into feat/issues/1579

SpencerTorres committed:
Update list

kavirajk committed:
Couple of additions

1. Populate column_gen.go
2. Removed unused the interface implementations

Signed-off-by: Kaviraj <kavirajkanagaraj@gmail.com>

kavirajk committed:
Merge branch 'main' into kavirajk/time-time64-support

kavirajk committed:
remove the need for license headers even in the template generators

Signed-off-by: Kaviraj <kavirajkanagaraj@gmail.com>

kavirajk committed:
Added following changes to time and time64 type

1. Removed the need for timezone information from these types (server
   explicitly states these types don't have timezone info)
2. Handle prefix correctly with time64(precision)
3. Handle default value for precision if not given
4. Update tests

Signed-off-by: Kaviraj <kavirajkanagaraj@gmail.com>

kavirajk committed:
Fix column generate for Time type

1. Also fixes the tests for timezone related tweaks

Signed-off-by: Kaviraj <kavirajkanagaraj@gmail.com>

kavirajk committed:
make sure `enable_time_time64_type` config is at right place

Clickhouse server returns error if client tries to make any queries with
unsupported settings.

In HTTP protocol, to find server version (ServerVersion() api) we were
using `enable_time_time64_type` as connection level settings and causing
the server to return error (any version before 25.6)

This PR don't pass any connection level settings to very the server
version and only if server version is > 25.6 it uses the settings to do
actual testing on time and time64 types

Signed-off-by: Kaviraj <kavirajkanagaraj@gmail.com>

kavirajk committed:
Fix some TYPES.md related docs and remarks

Signed-off-by: Kaviraj <kavirajkanagaraj@gmail.com>

SpencerTorres reviewed (changes requested):
Overall looks good, left some comments for things I had questions on. I understand this PR was copied from a different author. I appreciate your help ✨  enhancing ✨  it.

Let me know if you have any questions 👌

kavirajk committed:
PR remark: Fix column.tpl correctly

Time type doesn't need parse, because it doesn't take
any additional params

Signed-off-by: Kaviraj <kavirajkanagaraj@gmail.com>

kavirajk committed:
remove time.parse method

Signed-off-by: Kaviraj <kavirajkanagaraj@gmail.com>

kavirajk committed:
moving some repeated code to time helpers

Signed-off-by: Kaviraj <kavirajkanagaraj@gmail.com>

SpencerTorres commented on TYPES.md:
```diff
@@ -74,9 +74,53 @@ All types can be read into a pointer or pointer to a pointer.
+colTime64.col.WithPrecision(9)
+
+// Append values
+colTime.AppendRow(time.Date(2024, 7, 11, 12, 34, 56, 0, time.UTC))
```
This sounds much more correct, let's get ch-go updated too then. ch-go technically allows for these breaking changes with its versioning.

I'd say take all the code from here and ch-go and make it use `time.Duration`, it is much more fitting as you said.

kavirajk committed:
remove the need for custom serialization handling for time and time64

Signed-off-by: Kaviraj <kavirajkanagaraj@gmail.com>

kavirajk committed:
vendoring custom ch-go to have updated time and time64 types

Signed-off-by: Kaviraj <kavirajkanagaraj@gmail.com>

kavirajk committed:
Replace time.Time to time.Duration to work with Time and Time64

Changes:
1. Vendor the un-released version of ch-go to use the changes
2. Remove lots of helpers and parsing code that is unrelated to
   timezone, day, month and year
3. Update tests to validate the changes

Signed-off-by: Kaviraj <kavirajkanagaraj@gmail.com>

kavirajk committed:
Fix precision for time64 and split tests

Signed-off-by: Kaviraj <kavirajkanagaraj@gmail.com>

kavirajk committed:
Merge branch 'main' into kavirajk/time-time64-support

Signed-off-by: Kaviraj <kavirajkanagaraj@gmail.com>

kavirajk committed:
update docs in TYPES.md

Signed-off-by: Kaviraj <kavirajkanagaraj@gmail.com>

kavirajk commented on go.mod:
```diff
@@ -5,7 +5,7 @@ go 1.24.0
 
 require (
-	github.com/ClickHouse/ch-go v0.68.0
+	github.com/ClickHouse/ch-go v0.68.1-0.20251003160956-f6cfd63ac24c
```
FYI: This points to this [PR commit in ch-go.](https://github.com/ClickHouse/ch-go/pull/1097) Once it's merged we can point `main` (or after releasing point to specific newer version)

ernado replied:
Merged

kavirajk committed:
Merge branch 'main' into kavirajk/time-time64-support

Signed-off-by: Kaviraj <kavirajkanagaraj@gmail.com>

SpencerTorres reviewed (approved):
Looks usable, in a followup PR we should add tests for the stdlib. They're usually the same, but there's some quirks sometimes
